<h1 align=center><font size = 6>Segmenting and Clustering Neighborhoods in Toronto</font></h1>



## 1.Prepare the data

#### Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [8]:
!pip install wikipedia
import pandas as pd
import numpy as np
import wikipedia as wp
!pip install geocoder
import geocoder # import geocoder
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

!pip install folium
import folium # map rendering library

import requests
from pandas.io.json import json_normalize# tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#### Scrach the table from wikipedia via packages.

In [9]:
#Get the html source
html = wp.page("List_of_postal_codes_of_Canada:_M").html().encode("UTF-8")
df = pd.read_html(html)[0]

#### Select rows that have an assigned borough

In [10]:
df=df.loc[df['Borough']!='Not assigned',]

#### Assign the 'Borough' value to the 'Neiborhood' if the 'Neiborhood' under 'Not assigned'.

In [11]:
df.loc[df['Neighborhood']=='Not assigned','Neighborhood']=df['Borough']

#### Load CSV file

In [12]:
p_code=pd.read_csv('https://cocl.us/Geospatial_data')
p_code.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [13]:
df.shape[0]

103

# Check the joined table we got 103 rows and give its head.

In [14]:
df=df.join(p_code.set_index('Postal Code'),on='Postal Code')
print('The joined table have {} rows'.format(df.shape[0]))
df.head()

The joined table have 103 rows


Postal Code           Borough                                 Neighborhood  \
2         M3A        North York                                    Parkwoods   
3         M4A        North York                             Victoria Village   
4         M5A  Downtown Toronto                    Regent Park, Harbourfront   
5         M6A        North York             Lawrence Manor, Lawrence Heights   
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
2  43.753259 -79.329656  
3  43.725882 -79.315572  
4  43.654260 -79.360636  
5  43.718518 -79.464763  
6  43.662301 -79.389494

#### Get the location information of Toronto from geolocater

In [15]:

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a m|ap of New York with neighborhoods superimposed on top.

In [16]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'],df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Select rows where Borough contains 'Toronto'.

In [17]:
ind=df['Borough'].apply(lambda x: 1 if 'Toronto' in x else 0)
toronto_data=df.loc[ind==1,]
toronto_data.shape

(39, 5)

In [18]:
toronto_data.head()

Postal Code           Borough                                 Neighborhood  \
4          M5A  Downtown Toronto                    Regent Park, Harbourfront   
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
13         M5B  Downtown Toronto                     Garden District, Ryerson   
22         M5C  Downtown Toronto                               St. James Town   
30         M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  
4   43.654260 -79.360636  
6   43.662301 -79.389494  
13  43.657162 -79.378937  
22  43.651494 -79.375418  
30  43.676357 -79.293031

#### Define Foursquare Credentials and Version

## 2. Explore Neighborhoods in Manhattan

#### Define Foursquare Credentials and Version

In [19]:
CLIENT_ID = 'IH0PER1TCXPOAXOF0VWEBBKRZEUQIQJ2XURZVZ1ABPMC2QGN' # your Foursquare ID
CLIENT_SECRET = '0ZPQ5VDVU4MBO1LXLLQ1SMWJTBY2FQJEJPWUWCRQEQNZIRDG' # your Foursquare Secret
VERSION = '20200615' # Foursquare API version


#### Create a function to repeat the same process to all the neighborhoods in Toronto


In [20]:
def getNearbyVenues(names, latitudes, longitudes, radius=500,LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Catch venues of each neighboorhood and create a new dataframe called *Toronto_venues*.

In [21]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [22]:
print(toronto_venues.shape)
toronto_venues.head()

(1618, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

Let's check how many venues were returned for each neighborhood

In [23]:
toronto_venues.groupby('Neighborhood').count().head()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            56   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     18   
Central Bay Street                                                     65   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             56   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      18   
Central Bay Street                                                      65   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            56              56   
Brockton, Parkdale Village, Exhibition Place           23              23   
Business reply mail Processing Centre, South Ce...     16              16   
CN Tower, King and Spadina, Railway Lands, Harb...     18              18   
Central Bay Street                                     65              65   

                                                    Venue Longitude  \
Neighborhood                                                          
Berczy Park                                                      56   
Brockton, Parkdale Village, Exhibition Place                     23   
Business reply mail Processing Centre, South Ce...               16   
CN Tower, King and Spadina, Railway Lands, Harb...               18   
Central Bay Street                                               65   

                                                    Venue Category  
Neighborhood                                                        
Berczy Park                                                     56  
Brockton, Parkdale Village, Exhibition Place                    23  
Business reply mail Processing Centre, South Ce...              16  
CN Tower, King and Spadina, Railway Lands, Harb...              18  
Central Bay Street                                              65

#### Let's find out how many unique categories can be curated from all the returned venues

In [24]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 uniques categories.


## 3. Analyze Each Neighborhood

In [25]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Toy / Game Store  Trail  Train Station  \
0             0  ...                 0      0              0   
1             0  ...                 0      0              0   
2             0  ...                 0      0              0   
3             0  ...                 0      0              0   
4             0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  
0         0          0            0              0  
1         0          0            0              0  
2         0          0            0              0  
3         0          0            0              0  
4         0          0            0              0  

[5 rows x 237 columns]

And let's examine the new dataframe size.

In [26]:
toronto_onehot.shape

(1618, 237)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [27]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                                        Berczy Park     0.000000   
1       Brockton, Parkdale Village, Exhibition Place     0.000000   
2  Business reply mail Processing Centre, South C...     0.062500   
3  CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                 Central Bay Street     0.015385   

   Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0                0.0  0.000000            0.000000      0.000000   
1                0.0  0.000000            0.000000      0.000000   
2                0.0  0.000000            0.000000      0.000000   
3                0.0  0.055556            0.055556      0.055556   
4                0.0  0.000000            0.000000      0.000000   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0        0.000000         0.000000          0.000000                  0.0   
1        0.000000         0.000000          0.000000                  0.0   
2        0.000000         0.000000          0.000000                  0.0   
3        0.111111         0.166667          0.111111                  0.0   
4        0.000000         0.000000          0.000000                  0.0   

   ...  Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0  ...               0.0    0.0            0.0                       0.017857   
1  ...               0.0    0.0            0.0                       0.000000   
2  ...               0.0    0.0            0.0                       0.000000   
3  ...               0.0    0.0            0.0                       0.000000   
4  ...               0.0    0.0            0.0                       0.015385   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Wine Shop  Wings Joint  \
0               0.0                    0.0  0.000000        0.0          0.0   
1               0.0                    0.0  0.000000        0.0          0.0   
2               0.0                    0.0  0.000000        0.0          0.0   
3               0.0                    0.0  0.000000        0.0          0.0   
4               0.0                    0.0  0.015385        0.0          0.0   

   Women's Store  
0            0.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0  

[5 rows x 237 columns]

#### Let's confirm the new size

In [28]:
toronto_grouped.shape

(39, 237)

#### Let's print 3 neighborhoods along with the top 5 most common venues

In [29]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood'][0:3]:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2          Restaurant  0.04
3  Seafood Restaurant  0.04
4         Cheese Shop  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1     Coffee Shop  0.09
2  Breakfast Spot  0.09
3          Bakery  0.09
4             Gym  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0           Yoga Studio  0.06
1         Auto Workshop  0.06
2  Gym / Fitness Center  0.06
3         Garden Center  0.06
4                Garden  0.06




#### Sort the venues in descending order

In [30]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Create the new dataframe and display the top 10 venues for each neighborhood.

In [31]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...           Yoga Studio   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                Bakery                  Café   
1                Bakery           Coffee Shop        Breakfast Spot   
2         Auto Workshop                  Park           Pizza Place   
3      Airport Terminal        Airport Lounge              Boutique   
4                  Café    Italian Restaurant   Japanese Restaurant   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0           Cheese Shop              Beer Bar             Restaurant   
1         Grocery Store     Convenience Store  Performing Arts Venue   
2            Restaurant               Butcher          Burrito Place   
3   Rental Car Location                 Plane            Coffee Shop   
4        Sandwich Place          Burger Joint       Department Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Seafood Restaurant                   Pub               Creperie  
1             Pet Store          Climbing Gym             Restaurant  
2               Brewery            Skate Park     Light Rail Station  
3         Boat or Ferry       Harbor / Marina           Airport Gate  
4           Salad Place       Bubble Tea Shop            Yoga Studio

#### Built the K-Means model

In [32]:
# set number of clusters
kclusters = 6

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

#### Add labels to vanues data

In [33]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
4          M5A  Downtown Toronto                    Regent Park, Harbourfront   
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
13         M5B  Downtown Toronto                     Garden District, Ryerson   
22         M5C  Downtown Toronto                               St. James Town   
30         M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
4   43.654260 -79.360636               0           Coffee Shop   
6   43.662301 -79.389494               0           Coffee Shop   
13  43.657162 -79.378937               0        Clothing Store   
22  43.651494 -79.375418               0                  Café   
30  43.676357 -79.293031               0                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4                 Bakery                  Park                   Pub   
6       Sushi Restaurant           Yoga Studio                  Bank   
13           Coffee Shop       Bubble Tea Shop                  Café   
22           Coffee Shop          Cocktail Bar   American Restaurant   
30     Health Food Store                   Pub      Doner Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4             Restaurant        Breakfast Spot                  Café   
6               Beer Bar         Smoothie Shop        Sandwich Place   
13   Japanese Restaurant    Italian Restaurant        Cosmetics Shop   
22             Gastropub            Restaurant        Clothing Store   
30    Dim Sum Restaurant                 Diner        Discount Store   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4                     Theater                 Hotel           Dessert Shop  
6               Burrito Place                  Café     College Auditorium  
13  Middle Eastern Restaurant                 Diner                 Bakery  
22        Moroccan Restaurant        Cosmetics Shop               Creperie  
30        Distribution Center               Dog Run          Women's Store

##  4. Plot the clusters on map

In [34]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# IN CASE the notebook don't show the graph, here is the clusters screen capture

![cluster capture](https://github.com/DwayneLi/Coursera_Capstone/blob/master/WEEK3%20Question%203%20MAP.png?raw=true)